pandera/pandera.py

```diff
@@ -208,6 +209,128 @@ def __call__(self, parent_schema, check_index, check_obj):
+        Available relationships are: "greater_than", "less_than", "not_equal"
+    groupby : str|list[str]|callable|None
+        If a string or list of strings is
+        provided, then these columns are used to group the Column Series by
+        `groupby`. If a callable is passed, the expected signature is
+        DataFrame -> DataFrameGroupby. The function has access to the
+        entire dataframe, but the Column.name is selected from
+        this DataFrameGroupby object so that a SeriesGroupBy object is
+        passed into `fn`.
+
+        Specifying this argument changes the `fn` signature to:
+
+        dict[str|tuple[str], Series] -> bool|pd.Series[bool]
+
+        Where specific groups can be obtained from the input dict.
+
+    groups : str|list[str]|None
+        The dict input to the `fn` callable will be constrained to the
+        groups specified by `groups`.
+    test_kwargs : dict
+        Key Word arguments to be supplied to the test.
+    relationship_kwargs : dict
+        Key Word arguments to be supplied to the relationship function.
+        e.g. `alpha` could be used to specify a threshold in a t-test.
+    """
+    def __init__(self, test, relationship, groupby=None, groups=None,
+                 test_kwargs=None, relationship_kwargs=None):
+        #self.test = partial(test, **test_kwargs)
+        self.test = partial(test, **{} if test_kwargs is None else test_kwargs)
+        self.relationship = partial(self.relationships(relationship),
+                                    **relationship_kwargs)
+        self.groupby = groupby
+        self.groups = groups
+
+        super(Hypothesis, self).__init__(
+            self._check_fn, element_wise=False, groupby=groupby, groups=groups)
+
+    @staticmethod
+    def relationships(relationship):
+        # Impose a relationship on a supplied Test function.
+        if isinstance(relationship, str):
+            relationship = {
+                "greater_than": (lambda stat, pvalue, alpha:
+                                 stat > 0 and pvalue / 2 < alpha),
+                "less_than": (lambda stat, pvalue, alpha:
+                              stat < 0 and pvalue / 2 < alpha),
+                "not_equal": (lambda stat, pvalue, alpha:
+                              pvalue < alpha),
+            }[relationship]
+        elif not callable(relationship):
+            raise ValueError
+        return relationship
+
+    def _check_fn(self, check_obj):
+        # Creates a function fn which is checked via the Check parent class.
+        if self.groupby is None:
+            # one-sample case where no groupby argument supplied, apply to
+            # entire column
+            return self.relationship(*self.test(check_obj))
+        else:
+            return self.relationship(*self.test(*[check_obj.get(g)
+                                                  for g in self.groups]))
+
+    @classmethod
+    def two_sample_ttest(cls, groupby, groups, relationship, alpha=None,
```

Comment:
we should probably also check that `groups` is a list of exactly two strings.

failure tests for this would be to supply a string, or lists where len(groups) != 2.